pkg/service/storage/ucan/access_grant.go

```diff
@@ -0,0 +1,229 @@
+package ucan
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/multiformats/go-multihash"
+	"github.com/storacha/go-libstoracha/capabilities/access"
+	"github.com/storacha/go-libstoracha/capabilities/assert"
+	"github.com/storacha/go-libstoracha/capabilities/blob"
+	"github.com/storacha/go-libstoracha/capabilities/blob/replica"
+	"github.com/storacha/go-libstoracha/digestutil"
+	"github.com/storacha/go-ucanto/client"
+	"github.com/storacha/go-ucanto/core/dag/blockstore"
+	"github.com/storacha/go-ucanto/core/delegation"
+	"github.com/storacha/go-ucanto/core/invocation"
+	"github.com/storacha/go-ucanto/core/receipt/fx"
+	"github.com/storacha/go-ucanto/core/result"
+	"github.com/storacha/go-ucanto/core/result/failure"
+	"github.com/storacha/go-ucanto/principal"
+	"github.com/storacha/go-ucanto/server"
+	"github.com/storacha/go-ucanto/ucan"
+	"github.com/storacha/go-ucanto/validator"
+)
+
+const (
+	// UnknownAbilityErrorName is the name given to an error where the ability
+	// requested to be granted is unknown to the service.
+	UnknownAbilityErrorName = "UnknownAbility"
+	// UnknownCauseErrorName is the name given to an error where the cause
+	// invocation sent as context for the delegation is not recognised.
+	UnknownCauseErrorName = "UnknownCause"
+	// MissingCauseErrorName is the name given to an error where a required cause
+	// invocation has not been provided in the invocation to request a grant.
+	MissingCauseErrorName = "MissingCause"
+	// InvalidCauseErrorName is the name given to an error where the cause
+	// invocation has been determined to be invalid is some way. See the error
+	// message for details.
+	InvalidCauseErrorName = "InvalidCause"
+	// UnauthorizedCauseErrorName is the name given to an error where the cause
+	// invocation failed UCAN validation.
+	UnauthorizedCauseErrorName = "UnauthorizedCause"
+)
+
+// validity is the time a granted delegation is valid for.
+const validity = time.Hour
```

Comment:
Might be nice to plumb this into the configuration, though in practice seems fine. Maybe a TODO? Your call.

Reply:
I think we should be in charge of this value - i.e. don't let it be configured...so that we can change it later to get a smal perf bump if we wish.